### ISSUE TITLE
2.99 fails to index woowa-bm font files

### ISSUE THREAD

**Issue by no-preserve-root** [latex3/luaotfload#88]
Since 2.99, `luaotfload-tool` fails to index the woowa-bm font files, complaining about UTF-8 errors. The offending file is `BMDOHYEON_ttf.ttf` (although other files in the package might error out, too). The font is distributed [somewhere here](https://woowahan.com/#/fonts), although it might be easier to grab the file from the [ubuntu package](https://packages.ubuntu.com/disco/fonts/fonts-woowa-bm). Versions prior to 2.99 worked.

#### Steps to reproduce

1. Set up a fresh Ubuntu installation
2. Install TeXLive 2019
3. `apt update && apt install --no-install-recommends -y fonts-woowa-bm`
4. `luaoftload-tool -vvvvvvvvv -u`

#### Expected Result

Tool works.

#### Actual Result

[lots of unrelated output]
```
luaotfload | db : Loading font /usr/share/fonts/truetype/woowa-bm/BMJUA_ttf.ttf
luaotfload | db : Loading font /usr/share/fonts/truetype/woowa-bm/BMYEONSUNG_ttf.ttf
luaotfload | db : Loading font /usr/share/fonts/truetype/woowa-bm/BMDOHYEON_ttf.ttf
.../texmf-dist/tex/luatex/luaotfload/luaotfload-unicode.lua:55: invalid UTF-8 code
```

**Comment by zauguin** [comment]
@no-preserve-root Thanks for the report.

The problem is that we try to decode the Mac display name as Unicode string which fails, because it actually is encoded as Mac Roman. 

Given that this happens while we are trying to "sanitize" the fontname, i think we need two changes:

- [ ] Decode Mac Roman as Mac Roman and not as Unicode
- [ ] In case we encounter a Unicode name with invalid codepoints, skip them instead of printing an error. (That's basically the idea of this function in the first place.)

If this turns out to be difficult, revert https://github.com/latex3/luaotfload/pull/72 . The old code was broken too, but it was better hidden.

Off topic: The display name in question "should" decode to "πË¥ﬁ¿«πŒ¡∑ µµ«ˆ". I think this is a bug in the font.

**Comment by u-fischer** [comment]
Hm. Do we want to support fonts which encode names in some 8bit encoding? For me it sounds like a bug in the font.

**Comment by zauguin** [comment]
@u-fischer Generally OpenType allows multiple names in multiple languages, multiple encodings and for multiple platforms.
(E.g. IIRC Latin Modern has different names on Windows and Macintosh...)

This font for example contains a set of names for Windows (encoded in UTF-16) and a set of names for Macintosh (encoded in Mac Roman). In `luaotfload-database.lua` we prefer Macintosh names over Windows names because they have less problems. (I think Windows names tend to be changed for compatibility with weird, broken programs which are too expensive to replace, so they often do not have the specified structure.)

So *we* explicitly decide to use this oddly encoded names, which makes it hard to blame the font. (As I said above, the name also doesn't seem to be correctly encoded, but that's an independent issue.)

The problem is that the encoding information is discarded after the fontloader decodes the names for us but the fontloader only decodes the Windows names (UTF-16) and seems to assume that Mac names are either irrelevant or UTF-8 encoded.

**Comment by u-fischer** [comment]
Oh well. There one assume that the mac is utf8 and only windows has still problems with 8bit and then such a blow ;-). Does the reference to the fontloader mean that a "real" solution involves a change there too?

**Comment by zauguin** [comment]
Without patching the fontloader we can't reliably decode MacRoman because the encoding information is lost before we get it, but we could assume that all Macintosh names are MacRoman encoded (Given that we request english names that's a reasonable assumption), but should the fontloader add decoding for these names that would lead to problems. So I think proper decoding of MacRoman names has to wait until we make bigger rewrites of the database codes and rely less on the fontloader, especially since most MacRoman names turn out to be ASCII.

For the time being I pushed 57345f2522bfabe07f22cb36891ef1fde2bdd8a1 to our `dev` branch which restores the old behavior iff the name is not valid UTF-8.
I think that's a good solution at least for the second point from my first comment.

With this, indexing no longer fails.

**Comment by zauguin** [comment]
@u-fischer I *think* modern MacOS versions actually prefer the separate Unicode platform over the Mac platform and therefore actually use Unicode names, but there are no names in this font for the Unicode platform. (I'm also not sure if we would prefer Unicode or Macintosh)

**Comment by zauguin** [comment]
@u-fischer To correct myself, we actually prefer Windows over Macintosh and therefore normally get Unicode values, this is an exception because we also prefer the "compatiblefullname" over "fullname" and "compatiblefullname" only exists on Macintosh.

**Comment by u-fischer** [comment]
Hm. Travis fails, it can't find amiri after the change: \TU/Amiri(0)/m/n/10=Amiri:mode=node; at 10pt not loadable: metric data not found or bad. Locally it fails too. 

Beside this, it sounds as if one should release the change, so it should probably then be played back to master so that I can make a "2.991" from there.

**Comment by zauguin** [comment]
@u-fischer Travis should be happy now. The first version messed up the cache, so the fix might require cleaning the cache first off the broken version was used before.

I cherry picked both commits to master.

**Comment by u-fischer** [comment]
I adapted version etc. If travis is happy I will upload in the afternoon. Is there any sensible testfile possible for this?

**Comment by zauguin** [comment]
@u-fischer I can't think of any. We might want to add a OpenType font with outright invalid UTF-8 name fields to the tests to check if error handling works correctly.

**Comment by zauguin** [comment]
The new code in `dev` now tries to interpret everything which is not valid UTF-8 as Mac Roman. That should handle at least 99.99% of the cases.

**Comment by khaledhosny** [comment]
> This font for example contains a set of names for Windows (encoded in UTF-16) and a set of names for Macintosh (encoded in Mac Roman). In `luaotfload-database.lua` we prefer Macintosh names over Windows names because they have less problems.

I think the right fix here is to prefer the Unicode (in this case Windows) name entries. That is what many applications do, and they are the less problematic ones in my experience.

> (I think Windows names tend to be changed for compatibility with weird, broken programs which are too expensive to replace, so they often do not have the specified structure.)

There is [Typographic Family name](https://docs.microsoft.com/en-us/typography/opentype/spec/name#nid16) (id 16) and Typographic Subfamily name (id 17) that should be preferred over id 1 and 2 if present. The 1 and 2 ids are restricted to 4 fonts (regular, bold, italic, bold italic) per family for compatibility with old applications.

**Comment by zauguin** [comment]
@khaledhosny 

> I think the right fix here is to prefer the Unicode (in this case Windows) name entries. That is what many applications do, and they are the less problematic ones in my experience.

As I wrote above we *are* preferring Windows over Macintosh, the earlier comment got it mixed up. We still have to worry about encodings because we want to support fonts which do not contain Windows names.

> There is Typographic Family name (id 16) and Typographic Subfamily name (id 17) that should be preferred over id 1 and 2 if present. The 1 and 2 ids are restricted to 4 fonts (regular, bold, italic, bold italic) per family for compatibility with old applications.

We are using the typographic values by default, but this doesn't apply here because the name in question is the full name, not the family name.

**Comment by khaledhosny** [comment]
> @khaledhosny 
> We are using the typographic values by default, but this doesn't apply here because the name in question is the full name, not the family name.

Then `compatiblefullname` should not be given a priority, I don’t actually know any application that uses it.

**Comment by khaledhosny** [comment]
If full name is missing, it should be constructed from family name and style name, and `compatiblefullname` should be the last resort or not used at all (which is practically the same).

**Comment by zauguin** [comment]
@khaledhosny According to our documentation name based lookup should lookup based on the "menu name" of a font. So at least if Macintosh names are used and `compatiblefullname` is provided, then it is the right choice. We might improve the current situation by only using compatiblefullname if the fullname would be a Macintosh name otherwise. But even there I am not sure: For all fonts I found where "compatiblefullname" and "fullname" is provided, "compatiblefullname" is the better candidate. (Except for the font above, but there the "compatiblefullname" is just broken.)

**Comment by u-fischer** [comment]
@no-preserve-root Could you please report if the new version 2.991 on ctan resolved the problem?

**Comment by no-preserve-root** [comment]
@u-fischer The problem has been resolved with 2.991, thanks a lot!
